linera-views/src/journaling.rs

```diff
@@ -231,27 +249,81 @@ where
+    ///       or `key_len`. An upper bound is thus
+    ///       `journal_len_upper_bound = key_len + header_value_len`.
+    ///   Thus the following has to be taken as upper bound on block_size:
+    ///   `K::MAX_BATCH_TOTAL_SIZE - key_len - journal_len_upper_bound`.
+    ///
+    /// NOTE:
+    /// * Since a block must contain at least one operation and M bytes of the
+    /// serialization overhead (typically M = 2 or 3 bytes of vector sizes), condition (3)
+    /// requires that each operation in the original batch satisfies:
+    ///     `sizeof(key) + sizeof(value) + M <= K::MAX_VALUE_SIZE`
+    ///
+    /// * Similarly, a transaction must contain at least one block so it is desirable that
+    /// the maximum size of a block insertion `1 + sizeof(block_key) + K::MAX_VALUE_SIZE`
+    /// plus M bytes of overhead doesn't exceed the threshold of condition (2).
     async fn write_journal(
         &self,
-        simplified_batch: K::Batch,
+        batch: K::Batch,
         base_key: &[u8],
     ) -> Result<JournalHeader, K::Error> {
-        let mut iter = simplified_batch.into_iter();
+        let header_key = get_journaling_key(base_key, KeyTag::Journal as u8, 0)?;
+        let key_len = header_key.len();
+        let header_value_len = bcs::serialized_size(&JournalHeader::default())?;
+        let journal_len_upper_bound = key_len + header_value_len;
+        // Each block in a transaction comes with a key.
+        let max_transaction_size = K::MAX_BATCH_TOTAL_SIZE;
+        let max_block_size = std::cmp::min(
+            K::MAX_VALUE_SIZE,
+            K::MAX_BATCH_TOTAL_SIZE - key_len - journal_len_upper_bound,
+        );
+
+        let mut iter = batch.into_iter();
         let mut block_batch = K::Batch::default();
         let mut block_size = 0;
         let mut block_count = 0;
         let mut transaction_batch = K::Batch::default();
         let mut transaction_size = 0;
         while iter.write_next_value(&mut block_batch, &mut block_size)? {
             let (block_flush, transaction_flush) = {
-                if iter.is_empty() || block_batch.len() == K::MAX_BATCH_SIZE - 2 {
+                if iter.is_empty() || transaction_batch.len() == K::MAX_BATCH_SIZE - 1 {
                     (true, true)
                 } else {
-                    let next_block_size = iter.next_batch_size(&block_batch, block_size)?;
-                    let next_transaction_size = transaction_size + next_block_size;
-                    let transaction_flush = next_transaction_size > K::MAX_BATCH_TOTAL_SIZE;
-                    let block_flush = transaction_flush || next_block_size > K::MAX_VALUE_SIZE;
+                    // We can unwrap the result since if it were None, it would mean that
+                    // there is no next size but we have make sure this is not the case.
+                    let next_block_size = iter.next_batch_size(&block_batch, block_size)?.unwrap();
```

Comment:
No need for the lengthy comment. I would just write `.expect("iter is not empty")` which is clear from the condition above.

Reply:
Ok! Thank you for that.